### PR TITLE
Require at least a single match when defining discriminable properties

### LINKED ISSUE #53165: Intellisense suggests invalid property values within union objects
Does this issue occur when all extensions are disabled?: Yes

- VS Code Version: 1.76.0
- OS Version: Windows 11 22H2 (22621.1265)

Steps to Reproduce:

1. Using the following object union type:
```typescript
type TestType = {
  key1: 'a',
  key2: '1',
} | {
  key1: 'a',
  key2: '2',
} | {
  key1: 'b',
  key2: '3',
}
```
2. Even though `key2: '1' | '2'` when key1 is 'a', VSCode suggest 1, 2 and 3 as b values:
image
The underlying type checker even knows the correct values for key2, and throws a problem if theyre used: (note the `key2: "1" | "2"` part)
image

This causes bigger usability problems when used with larger unions and objects where VSCode spills the suggestions with lots of incorrect ones.

### PR DESCRIPTION
**Pull request by MariaSolOs**
Fixes #53165

### CODE REVIEW

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -29246,6 +29247,7 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
     function isPossiblyDiscriminantValue(node: Expression): boolean {
         switch (node.kind) {
             case SyntaxKind.StringLiteral:
+                return !isEmptyStringLiteral(node as StringLiteral);
```
Comment (Andarist) [inline src/compiler/checker.ts:29250]: I think that it's likely that we could fix this in a more general way if we'd do **something** like this here:
```suggestion
                return !hasSkipDirectInferenceFlag(node);
```

The problem is that only nodes within a `containingCall` are "blocked"/skipped today:
https://github.dev/microsoft/TypeScript/blob/a280cafbf85df16bcd2f7258f26fdb85615b3c41/src/compiler/checker.ts#L1851-L1858

Maybe it's just a matter of adding a new stop condition to this ancestry traversal?

Comment (Andarist) [reply src/compiler/checker.ts:29250]: Yeah, I couldn't resist tinkering with this a little bit more and I think that this kind of solution is on the right track.

With this patch the expanded test case works:
```diff
-        const containingCall = findAncestor(node, isCallLikeExpression);
+        const containingCall = findAncestor(node, n => isCallLikeExpression(n) || !!n.parent && isVariableDeclaration(n.parent));
```

I used this condition only as a quick test to see if it could work. By no means it's correct/complete. I think that we are looking for a parent that "stops" providing contextual types as we should mark everything up until that parent. Since contextual types come from lots of places... I'm not sure how to write such a condition on the spot :p

File: tests/cases/fourslash/completionsUnionProperty.ts [hunk 2]
```diff
@@ -0,0 +1,6 @@
+/// <reference path="fourslash.ts" />
+
+//// type Foo = { a: 0, b: 'x' } | { a: 0, b: 'y' } | { a: 1, b: 'z' };
+//// const foo: Foo = { a: 0, b: '/**/
+
+verify.completions({ marker: "", triggerCharacter: "'", includes: [ "x", "y" ], excludes: [ "z" ] });
```
[the remark this answers is not in the document]
Comment (MariaSolOs) [reply tests/cases/fourslash/completionsUnionProperty.ts:6]: @Andarist you're right. I was looking back at this proposed solution and it doesn't feel right. Will continue working on it! :)

File: src/compiler/checker.ts (hunk 1 above)
Comment (Andarist) [reply src/compiler/checker.ts:29250]: > @Andarist the idea I have in mind is doing a smarter filtering [here](https://insiders.vscode.dev/github/MariaSolOs/TypeScript/blob/2f75e38bc40d0c58ab74ba8cd4ba0dcad33ce4f3/src/compiler/checker.ts#L29271-L29275) and filter out properties where the initializer doesn't match the property's type.

That makes a lot of sense 👍

File: tests/cases/fourslash/completionsUnionProperty.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (MariaSolOs) [reply tests/cases/fourslash/completionsUnionProperty.ts]: Okay I included this test and I think I came up with a more generic solution.

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -22628,11 +22628,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let i = 0;
             for (const type of target.types) {
                 const targetType = getTypeOfPropertyOfType(type, propertyName);
-                if (targetType && related(getDiscriminatingType(), targetType)) {
-                    discriminable[i] = discriminable[i] === undefined ? true : discriminable[i];
-                }
-                else {
-                    discriminable[i] = false;
+                if (discriminable[i] === undefined) {
```
Comment (MariaSolOs) [inline src/compiler/checker.ts:22631]: To explain this change: [This line](https://insiders.vscode.dev/github/MariaSolOs/TypeScript/blob/86295fb3d7ffa184f2f00f539171c56c56826961/src/compiler/checker.ts#L22621-L22622) says that "discriminated types stop at 'true'", but the code is overwriting a potentially true `discriminable[i]` to `false` if one of the union variants has a property with a type that's unrelated to the target type.

File: tests/baselines/reference/assignmentCompatWithDiscriminatedUnion.errors.txt [hunk 4]
```diff
@@ -6,10 +6,6 @@ tests/cases/conformance/types/typeRelationships/assignmentCompatibility/assignme
 tests/cases/conformance/types/typeRelationships/assignmentCompatibility/assignmentCompatWithDiscriminatedUnion.ts(58,5): error TS2322: Type 'S' is not assignable to type 'T'.
   Property 'c' is missing in type 'S' but required in type '{ a: 2; b: 4 | 3; c: string; }'.
 tests/cases/conformance/types/typeRelationships/assignmentCompatibility/assignmentCompatWithDiscriminatedUnion.ts(82,5): error TS2322: Type 'S' is not assignable to type 'T'.
-  Type 'S' is not assignable to type '{ a: N; b: N; c: 2; }'.
-    Types of property 'c' are incompatible.
-      Type 'N' is not assignable to type '2'.
-        Type '0' is not assignable to type '2'.
```
Comment (MariaSolOs) [inline tests/baselines/reference/assignmentCompatWithDiscriminatedUnion.errors.txt:12]: I'm not a fan of these errors not being included anymore and I'll try to keep them, but IMO the diagnostic changes in the other files do make sense.

**Comment by RyanCavanaugh** [comment]
@typescript-bot test this
@typescript-bot test top100
@typescript-bot user test this
@typescript-bot user test tsserver
@typescript-bot test tsserver top100
@typescript-bot run dt
@typescript-bot perf test this

**Comment by MariaSolOs** [comment]
> @RyanCavanaugh Ryan Cavanaugh FTE Here are the results of running the top-repos suite comparing `main` and `refs/pull/53654/merge`:
> 
> Something interesting changed - please have a look.
> 
> Details

Ehhh this looks unrelated to this change... right?

**Comment by DanielRosenwasser** [comment]
> Ehhh this looks unrelated to this change... right?

I *think* so.

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -22628,11 +22628,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let i = 0;
             for (const type of target.types) {
                 const targetType = getTypeOfPropertyOfType(type, propertyName);
-                if (targetType && related(getDiscriminatingType(), targetType)) {
-                    discriminable[i] = discriminable[i] === undefined ? true : discriminable[i];
-                }
-                else {
-                    discriminable[i] = false;
+                if (discriminable[i] === undefined) {
+                    discriminable[i] = !!(targetType && related(getDiscriminatingType(), targetType));
```
Comment (DanielRosenwasser) [inline src/compiler/checker.ts:22632]: What happens if you switch this to check `discriminable[i] !== true` so that as long as *some* type has a successful discriminant, the property is discriminable?

Or alternatively,

```ts
checkEachType: for (const type of target.types) {
    const targetType = getTypeOfPropertyOfType(type, propertyName);

    const targetIsDiscriminable = discriminable[i] = !!(targetType && related(getDiscriminatingType(), targetType));
    if (targetIsDiscriminable) {
        break checkEachType;
    }

    i++
}
```

Comment (MariaSolOs) [reply src/compiler/checker.ts:22632]: That doesn't change the logic though, it just optimizes the looping a bit.

Comment (MariaSolOs) [reply src/compiler/checker.ts:22632]: I do like your suggestion BTW, just making sure I'm not missing something.

Comment (DanielRosenwasser) [reply src/compiler/checker.ts:22632]: Doesn't it?

In the original version, a single `false` takes precedence over everything else.

With your change, the first result takes precedence over every other (which sounds suspicious - this doesn't even need to be a loop anymore).

In my suggestion, a single `true` takes precedence over any other result.

Comment (MariaSolOs) [reply src/compiler/checker.ts:22632]: Right, that makes sense. Boolean logic gets tricky with truthy and falsy stuff.

Comment (MariaSolOs) [reply src/compiler/checker.ts:22632]: Also, with my change the first result doesn't always take precedence, because the objects in the union could potentially contain different items and so the writes to `discriminable` could differ between iterations.

Comment (MariaSolOs) [reply src/compiler/checker.ts:22632]: @DanielRosenwasser your suggested change breaks [this test](https://github.com/microsoft/TypeScript/blob/e83d61398ea0e4231e882121dd6c6bcfe4fdc9e4/tests/cases/compiler/contextualTypeShouldBeLiteral.ts#L40). 
```
foo2({
  type2: 'y',
  ~~~~~
!!! error TS2345: Argument of type '{ type2: string; value: string; method(): void; }' is not assignable to parameter of type 'X2 | Y2'.
!!! error TS2345:   Object literal may only specify known properties, but 'type2' does not exist in type 'X2'. Did you mean to write 'type1'?
  value: 'done',
  method() {
    this;
    this.value;
  }
});
```

File: tests/baselines/reference/assignmentCompatWithDiscriminatedUnion.errors.txt (hunk 4 above)
Comment (MariaSolOs) [reply tests/baselines/reference/assignmentCompatWithDiscriminatedUnion.errors.txt:12]: @DanielRosenwasser @ahejlsberg Wait I actually think that those errors are a bug (?). The example says that `s` should be assignable to `t` (and if you look at it, it is), but it errors because of the number of combinations being [too complex](https://github.com/microsoft/TypeScript/blob/68d8be4f1d89d94a543d93fcedad17f3e4ef297b/tests/cases/conformance/types/typeRelationships/assignmentCompatibility/assignmentCompatWithDiscriminatedUnion.ts#L63-L82). The diagnostics are incorrect, and so I'm actually prefer them going away.

**Review by DanielRosenwasser — APPROVED** [review]
I have some other ideas for error recovery here, but I *believe* this is an improvement that doesn't have to wait on that.

File: src/compiler/checker.ts [hunk 6]
```diff
@@ -22628,11 +22628,8 @@ export function createTypeChecker(host: TypeCheckerHost): TypeChecker {
             let i = 0;
             for (const type of target.types) {
                 const targetType = getTypeOfPropertyOfType(type, propertyName);
```
Comment (jakebailey) [inline src/compiler/checker.ts:22630]: Can this calculation be inside the loop so we aren't doing any work if the i-th element has already been calculated?

Comment (MariaSolOs) [reply src/compiler/checker.ts:22630]: Good point.

Comment (jakebailey) [reply src/compiler/checker.ts:22630]: Though, it could be a behavior change... Doubt it, though.

**Comment by jakebailey** [comment]
@typescript-bot perf test this faster

File: src/compiler/checker.ts (hunk 6 above)
Comment (MariaSolOs) [reply src/compiler/checker.ts:22630]: If it changes stuff, then I truly know nothing about booleans and loops lol.

Comment (jakebailey) [reply src/compiler/checker.ts:22630]: Usually it comes down to "will this call check something that causes an error that hasn't previously been seen before?".

**Comment by ahejlsberg** [comment]
The fix in this PR unfortunately doesn't work right in all cases. Consider:

```ts
type Foo = { x: 0, y: 0, z: 'a' } | { x: 0, y: 1, z: 'b' } | { x: 1, y: 0, z: 'c' } | { x: 1, y: 1, z: 'd' };

let foo: Foo = {
    x: 0,
    y: 1,
    z: // Suggests "a" | "b" here
}
```

Above, only `"b"` should be in the completion list. The issue is that once `discriminable[i]` gets set to `true` for a particular constituent, it isn't possible for a later check to disqualify that constituent. So it doesn't work when there are multiple discriminant properties as in the example above.

In general, I find the original logic in `discriminateTypeByDiscriminableItems` hard to follow, and I can't really convince myself that it is correct. It basically starts out by saying that _no_ constituents are in the discriminated result and then only includes constituents when they have a discriminant property with a matching type. I think it needs to be the other way around: By default, _all_ constituents should be in the discriminated result, but when constituents have discriminant properties with types that _don't_ match, they should be removed from the result. Then, in order to fix #53165, we need to additionally say that we only remove constituents when some other constituent has a discriminant property with a matching type (in other words, if we know a property will for sure cause an error, we should just ignore that property).

I'm going to put up a PR with the suggested changes.

**Comment by MariaSolOs** [comment]
@ahejlsberg oh well, I tried 😅 Thank you for coming up with a proper solution!

**Comment by MariaSolOs** [comment]
Closing as this PR has been superseded by #53709
